Doc/regions.md

```diff
@@ -0,0 +1,517 @@
+```python
+def closure3():
+  x = 1
+  def inc():
+    nonlocal x
+    x += 1
+    return x
+  def dec():
+    nonlocal x
+    x -= 1
+    return x
+  return inc, dec
+```
+This returns a tuple of two function objects that share the same state.
+The first function increments the value of `x`, and the second decrements it.
+```python
+>>> inc, dec = closure3()
+>>> inc()
+2
+>>> dec()
+1
+```
+
+Under the hood in the CPython implementation, closures are implemented using a `Cell` object.
+A `Cell` object is a simple object that contains a single field.
+When a closure is created, the variables that are captured are stored in `Cell` objects.
+
+```Python
+>>>  inc.__closure__
+(<cell at 0x7ea265cdb490: int object at 0x7ea268160110>,)
+>>>  dec.__closure__
+(<cell at 0x7ea265cdb490: int object at 0x7ea268160110>,)
+```
+Here we can see that the two function objects have closed over the same `cell` object.
+
+The Python interpreter conducts a free variable analysis and decides if a variable is used in a function object,
+and if such it creates a `Cell` object to store the value of the variable.
+
+### `locals()`
+
+In Python, it is possible to create an object that contains all the local variables in the current scope.
+
+```python
+def locals_example1():
+  x = 1
+  l = locals()
+  return l
+```
+This function returns a dictionary object that contains all the local variables in the current scope.
+```python
+>>> locals_example1()
+{'x': 1}
+```
+
+The `locals()` function returns a dictionary object that contains all the local variables in the current scope.
+This dictionary object is a snapshot of the local variables at the time the function is called.
+It is not a live view of the local variables.
+```python
+def locals_example2():
+  x = 1
+  l = locals()
+  x = 2
+  return l
+```
+Calling this function the returned dictionary does not account for the update to `x`.
+```python
+>>> locals_example2()
+{'x': 1}
+````
+
+The `locals()` function also captures the local variables in the enclosing scope.
+```python
+def locals_example3():
+  x = 1
+  y = 2
+  def inner():
+    t = x
+    return locals()
+  return inner()
+```
+This returns a dictionary object that contains the local variables in the enclosing scope that have been mentioned.
+```python
+>>> locals_example3()
+{'t': 1, 'x': 1}
+```
+The variable `x` has been captured by the `locals()` function,
+but `y` has not been mentioned in the inner function, and thus is not captured.
+
+The capture is also nested. 
+```python
+def locals_example4():
+  x = 1
+  y = 2
+  def inner():
+    def inner_inner():
+      t = x
+    return locals()
+  return inner()
+```
+Here the dictionary returned `locals` contains the function object `inner_inner` and `x`.
+The `t` used in `inner_inner` is not captured as it is in an inner scope.
+```python
+>>> locals_example4()
+{'inner_inner': <function locals_example4.<locals>.inner.<locals>.inner_inner at 0x7ea265d3add0>, 'x': 1}
```

Comment:
Why is `x` in `locals()` here? Is it because it was mentioned in `inner_inner()`?